staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/controller.go

```diff
@@ -0,0 +1,420 @@
+		allBindings := definitionInfo.dependentBindings
+
+		if !ok || len(allBindings) == 0 {
+			// Definition not yet reconciled? No bindings for this def?
+			continue
+		}
+
+		for bindingName := range allBindings {
+			bindingNamespace, bindingName, err := cache.SplitMetaNamespaceKey(bindingName)
+			if err != nil {
+				// programming/internal error. these should all be parsable
+				utilruntime.HandleError(fmt.Errorf("failed to parse key %v: %w", bindingName, err))
+				continue
+			}
+
+			var binding fantasyInternalPolicyBinding
+			if len(bindingNamespace) == 0 {
+				binding, err = c.policyBindingController.
+					Informer().
+					Get(bindingName)
+			} else {
+				binding, err = c.policyBindingController.
+					Informer().
+					Namespaced(bindingNamespace).
+					Get(bindingName)
+			}
+
+			if err != nil {
+				if k8serrors.IsNotFound(err) {
+					// Binding not found from list. Clean it up
+					allBindings.Delete(bindingName)
+					continue
+				} else {
+					// Other, bad error
+					utilruntime.HandleError(fmt.Errorf("cel-admission-controller: failed to look up binding %w", err))
+					continue
+				}
+			}
+
+			if !binding.Matches(a) {
+				continue
+			}
+
+			bindingInfo := c.bindingInfos[bindingName]
+			if bindingInfo.evaluator == nil {
+				// Compile policy definition using binding
+				//!TODO: for each binding should there be an evaluator for each
+				// target GVK?
+				bindingInfo.evaluator, err = definition.Compile(nil)
+				if err != nil {
+					// compilation error. What to do with it?
+					utilruntime.HandleError(fmt.Errorf("failed to compile CEL expression: %w", err))
+					continue
+				}
+				c.bindingInfos[bindingName] = bindingInfo
+			}
+
+			// Fetch params object by indexing
+			paramsGr := definition.ParamSource()
+			paramsNamespace, paramsName := binding.TargetParams()
+			paramInfo, ok := c.paramsCRDControllers[paramsGr]
+			if !ok {
+				// Not watching this binding's params CRD? seems like a bug yo
+				continue
+			}
+
+			var param *unstructured.Unstructured
+			if len(paramsNamespace) == 0 {
+				param, err = paramInfo.controller.Informer().Get(paramsName)
+			} else {
+				param, err = paramInfo.controller.Informer().Namespaced(paramsNamespace).Get(paramsName)
+			}
+
+			if err != nil {
+				if k8serrors.IsNotFound(err) {
+					// Param doesnt exist yet?
+					// Maybe just have to wait a bit.
+					continue
+				}
+
+				// There was a bad internal error
+				utilruntime.HandleError(err)
+				continue
+			}
+
+			decisions := bindingInfo.evaluator(a, param)
+			for _, decision := range decisions {
+				switch decision.PolicyDecisionKind {
+				case Admit:
+					// Do nothing
+				case Warn:
+					// How to surface warning to user? Just log in apiserver logs?
+				case Deny:
+					allDecisions = append(allDecisions, decision)
+				default:
+					// unrecognized decision. ignore
+				}
+			}
+		}
+	}
+
+	if len(allDecisions) > 0 {
+		return errors.New("deny")
```

Comment:
Maybe future work, but we want to ultimately be able to respond with message(s), and a status and code here like webhooks do. The response back to the client request (the one being validated for admission) should use the status and code.

Reply:
This brings up a question I didn't fully cover in the KEP:

When multiple policy definition validations deny admission, which status and code are returned? The alternatives I'm aware of:

1. Agree on an order of precedent of http response codes, use the "most severe" code
2. Use the code of whatever policy fails first

If we go with (2) we need to make sure we validate policies (and bindings) in a consistent order.  But I would love to see (1) work, I just haven't thought it through carefully enough to know if there is a sane ordering.

Reply:
Since we only return one status/code back, do we include all validation failures in messages?

Reply:
I was assuming if, for any reason, any validation is a `Deny`, then the response code is unconditionally `409 Conflict`. Should policy definitions be able to control the HTTP status code? Are there other codes that one might choose?

Re: returning multiple errors. I believe we can return a JSON-formatted error string with some metadata like web hooks do. In this case we can return an JSON array with an entry for each validation that failed where the requestor can look for more information.